src/com/amazon/ion/impl/bin/PooledBlockAllocatorProvider.java

```diff
@@ -76,13 +77,24 @@ public int getBlockSize()
         public void close() {}
     }
 
+    // A globally shared instance of the PooledBlockAllocatorProvider that is lazily initialized.
+    // This instance allows BlockAllocators to be re-used across instantiations of classes like
+    // the binary Ion writer, thereby avoiding costly array initializations.
+    private static PooledBlockAllocatorProvider instance;
     private final ConcurrentMap<Integer, BlockAllocator> allocators;
 
-    public PooledBlockAllocatorProvider()
+    private PooledBlockAllocatorProvider()
     {
         allocators = new ConcurrentHashMap<Integer, BlockAllocator>();
     }
 
+    public static synchronized PooledBlockAllocatorProvider getInstance() {
+        if (instance == null) {
+            instance = new PooledBlockAllocatorProvider();
+        }
+        return instance;
+    }
```

Comment:
Can we put this behind static initializer (i.e. at class load time) versus using a mutex and check code?

Reply:
If you want a static initializer there's always the [`enum` trick](https://stackoverflow.com/questions/70689/what-is-an-efficient-way-to-implement-a-singleton-pattern-in-java). Is there some reason not to do that?

Reply:
I am more raising my eyebrows at the `getInstance()` behind a mutex. This code *should* work as a part of class initialization and you get the memory barrier there, so I am more wondering why not just say `private final static INSTANCE = new PooledBlockAllocatorProvider();` or the like.

That said, if this is required for some reason, let's comment why.

Reply:
Yeah, that makes sense.